docs/architecture.md

```diff
@@ -1,32 +1,139 @@
+
+All distributors share access to the same hash ring, which means that write requests can be sent to any distributor.
+
+#### Quorum consistency
+
+To ensure consistent query results, Cortex uses DynamoDB-style quorum consistency on reads and writes. This means that the distributor will wait for a positive response of at least one half plus one of the ingesters to send the sample to before responding to the user.
+
+#### Load balancing across distributors
+
+We recommend randomly load balancing write requests across distributor instances, potentially by running the distributors as a Kubernetes [Service](https://kubernetes.io/docs/concepts/services-networking/service/).
+
+### Ingester
+
+The **ingester** service is responsible for writing sample data to long-term storage backends (DynamoDB, S3, Cassandra, etc.).
+
+#### Shared-nothing processes
+
+Ingesters are, for the most part, shared-nothing processes. They don't coordinate or communicate with each other *under normal operation*. The exception is on rolling updates, during which ingesters transfer their in-memory data from the ingester exiting the group to the ingester joining the group.
+
+#### State
+
+Ingesters are semi-stateful in that they always retain the last 12 hours worth of samples. When restarting or upgrading ingesters, care must be taken to avoid losing that data.
+
+As *semi*-stateful processes, ingesters are *not* designed to be long-term data stores. In Cortex, that role is played by the [chunk store](#chunk-store).
+
+#### Write de-amplification
+
+Ingesters semi-statefully store the last 12 hours worth of samples in order to perform **write de-amplification**, i.e. batching and compressing samples for the same series and flushing them out to the [chunk store](#chunk-store). Under normal operations, there should be *many* orders of magnitude fewer queries per second (QPS) worth of writes to the chunk store than to the ingesters.
+
+This division of labor between the ingester and the chunk store is the main source of Cortex's low total cost of ownership (TCO).
+
+### Querier
+
+The **querier** service handles the actual [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/) evaluation of samples stored in long-term storage.
+
+It embeds the chunk store client code for fetching data from long-term storage and communicates with [ingesters](#ingester) for more recent data.
+
+### Ruler
+
+The **ruler** service is responsible for handling alerts produced by [Alertmanager](https://prometheus.io/docs/alerting/alertmanager/).
+
+### Query frontend
+
+The **query frontend** is an optional service that accepts HTTP requests, queues them by tenant ID, and retries in case of errors.
+
+> As mentioned [above](#distributor), the query frontend is completely optional; you can use queriers directly. To use the query frontend, direct incoming authenticated traffic at them and set the `-querier.frontend-address` flag on the queriers.
```

Comment:
Drop "As mentioned [above](#distributor)," as it shouldn't be mentioned there, and make it clear these are for queries (reads).